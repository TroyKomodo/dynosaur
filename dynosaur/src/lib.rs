#![doc = include_str!("../README.md")]

pub use dynosaur_derive::*;
