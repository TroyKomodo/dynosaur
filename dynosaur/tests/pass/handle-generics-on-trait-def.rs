use dynosaur::dynosaur;

#[dynosaur(DynMyTrait)]
trait MyTrait<T> {
    const N: i32;
    type Item;
    async fn foo(&self, x: &T) -> i32;
}

fn main() {}
