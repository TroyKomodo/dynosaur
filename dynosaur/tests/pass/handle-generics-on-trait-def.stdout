#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait<T> {
    const N: i32;
    type Item;
    async fn foo(&self, x: &T)
    -> i32;
}
trait ErasedMyTrait<T> {
    const N: i32;
    type Item;
    fn foo<'life0, 'life1, 'dynosaur>(&'life0 self, x: &'life1 T)
    ->
        ::core::pin::Pin<Box<dyn ::core::future::Future<Output = i32> +
        'dynosaur>>
    where
    'life0: 'dynosaur,
    'life1: 'dynosaur,
    Self: 'dynosaur;
}
impl<T, DYNOSAUR: MyTrait<T>> ErasedMyTrait<T> for DYNOSAUR {
    const N: i32 = <Self as MyTrait<T>>::N;
    type Item = <Self as MyTrait<T>>::Item;
    fn foo<'life0, 'life1, 'dynosaur>(&'life0 self, x: &'life1 T)
        ->
            ::core::pin::Pin<Box<dyn ::core::future::Future<Output = i32> +
            'dynosaur>> where 'life0: 'dynosaur, 'life1: 'dynosaur,
        Self: 'dynosaur {
        Box::pin(<Self as MyTrait<T>>::foo(self, x))
    }
}
struct DynMyTrait<'dynosaur, T, Item> {
    ptr: *mut (dyn ErasedMyTrait<T, Item = Item> + 'dynosaur),
    owned: bool,
}

fn main() {}
